File: src/nav.cpp
#include <ros/ros.h>
#include <geometry_msgs/PoseArray.h>
#include <geometry_msgs/PoseStamped.h>
#include <actionlib/client/simple_action_client.h>
#include <move_base_msgs/MoveBaseGoal.h>
#include <move_base_msgs/MoveBaseAction.h>
#include <std_msgs/Bool.h>
#include <list>

actionlib::SimpleActionClient<move_base_msgs::MoveBaseAction>* client_ptr;
std::list<geometry_msgs::Pose> poses;

void navGoalCb(const geometry_msgs::PoseArray::ConstPtr& msg)
{
    poses.clear();
    ROS_INFO_STREAM("NEW NAV GOAL!");
    for(geometry_msgs::Pose p : msg->poses)
    {
	poses.push_back(p);
    }
}

void confirmationCB(const std_msgs::Bool::ConstPtr& msg)
{
    if(msg->data){
        if(poses.size() <= 0){
            ROS_INFO_STREAM("Plan is size 0!");
            return;
        }
        ROS_INFO_STREAM("Executing next trajectory!");
        move_base_msgs::MoveBaseGoal goal;
        goal.target_pose.pose = poses.front();
	goal.target_pose.header.frame_id = "map";
        client_ptr->sendGoal(goal);

        poses.pop_front();
    }
}

int main(int argc, char** argv){
    ros::init(argc, argv, "nav_node");
    ros::NodeHandle nh;

    actionlib::SimpleActionClient<move_base_msgs::MoveBaseAction> client("move_base");
    client_ptr = &client;
    ROS_INFO_STREAM("STARTING NAV NODE!");

    ros::Subscriber goal_sub = nh.subscribe("/navigation_goal", 1000, navGoalCb);
    ros::Subscriber confirmation_sub = nh.subscribe("/nav_confirmation", 1000, confirmationCB);

    ros::spin();

    return 0;
}

File: src/nav.py
#! /usr/bin/env python2
import rospy
from std_msgs.msg import Bool
from geometry_msgs.msg import PoseArray
from tf2_geometry_msgs import PoseStamped
from geometry_msgs.msg import PoseStamped as GeoPoseStamped
import tf2_ros
import actionlib


class Nav:
    def __init__(self):
        self.poses = []
        self.result_pub = rospy.Publisher("/nav_goal_pose", GeoPoseStamped, queue_size=10)
        rospy.Subscriber("/nav_confirmation", Bool, self.confirmation_cb)
        rospy.Subscriber("/navigation_goal", PoseArray, self.nav_goal_cb)
        self.tf_buffer = tf2_ros.Buffer()
        self.tf_listener = tf2_ros.TransformListener(self.tf_buffer)

    def nav_goal_cb(self, msg):
        # unpack the pose array, add each pose to the array poses
        rospy.loginfo("[SPOT_NAV]: Recieved nav goal pose array of length {}".format(len(msg.poses)))
        self.poses = msg.poses

    def confirmation_cb(self, msg):
        if msg.data:
            # pop front pose off of poses
            if len(self.poses) <= 0:
                rospy.logerr("[SPOT_NAV]: Confirmation sent but no goal poses currently queued")
                return
            pose = self.poses.pop(0)


            ps = PoseStamped()
            ps.pose = pose
            ps.header.frame_id = "odom"
            ps = self.tf_buffer.transform(ps, "body", rospy.Duration(1))
            gps = GeoPoseStamped()
            gps.header = ps.header
            gps.pose = ps.pose
            self.result_pub.publish(gps)


if __name__ == "__main__":
    rospy.init_node('fetch_backend')
    spotNav = Nav()
    rospy.loginfo("[SPOT_NAV]: Backend up and running!")
    rospy.spin()
